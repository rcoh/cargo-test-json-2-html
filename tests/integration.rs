use cargo_test_json_2_html::{convert_to_html, Config};

#[test]
fn test_real_cargo_output() {
    let input = r#"{ "type": "suite", "event": "started", "test_count": 3 }
{ "type": "test", "event": "started", "name": "tests::test_fail" }
{ "type": "test", "event": "started", "name": "tests::test_ignored" }
{ "type": "test", "event": "started", "name": "tests::test_pass" }
{ "type": "test", "name": "tests::test_ignored", "event": "ignored" }
{ "type": "test", "name": "tests::test_pass", "event": "ok", "stdout": "This test passes with stdout\nThis test has stderr output\n" }
{ "type": "test", "name": "tests::test_fail", "event": "failed", "stdout": "This test fails with stdout\nThis test has stderr output\n\nthread 'tests::test_fail' panicked at src/main.rs:18:9:\nassertion `left == right` failed\n  left: 4\n right: 5\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n" }
{ "type": "suite", "event": "failed", "passed": 1, "failed": 1, "ignored": 1, "measured": 0, "filtered_out": 0, "exec_time": 0.000290792 }"#;

    let config = Config::default();
    let html = convert_to_html(input, config);
    
    insta::assert_snapshot!(html);
}

#[test]
fn test_mixed_output_with_compilation() {
    let input = r#"   Compiling cargo-test-json-2-html v0.1.0 (/Users/rcoh/cargo-test-json-2-html)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.17s
     Running unittests src/main.rs (target/debug/deps/cargo_test_json_2_html-f06c6ae57aba77b0)

{ "type": "suite", "event": "started", "test_count": 1 }
{ "type": "test", "event": "started", "name": "tests::simple_test" }
{ "type": "test", "name": "tests::simple_test", "event": "ok", "stdout": "test output\n" }
{ "type": "suite", "event": "ok", "passed": 1, "failed": 0, "ignored": 0, "measured": 0, "filtered_out": 0, "exec_time": 0.001 }"#;

    let config = Config::default();
    let html = convert_to_html(input, config);
    
    insta::assert_snapshot!(html);
}

#[test]
fn test_error_handling() {
    let input = r#"{ "type": "test", "invalid": "json" }
{ "type": "test", "name": "valid_test", "event": "ok" }
{ not json at all
Some compilation output"#;

    let config = Config::default();
    let html = convert_to_html(input, config);
    
    insta::assert_snapshot!(html);
}
