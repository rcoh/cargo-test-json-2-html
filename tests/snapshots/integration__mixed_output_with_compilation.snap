---
source: tests/integration.rs
expression: html
---
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Cargo Test Report</title>
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 20px; background: #f5f5f5; }
        .container { max-width: 1200px; margin: 0 auto; background: white; border-radius: 8px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
        .header { background: #2563eb; color: white; padding: 20px; border-radius: 8px 8px 0 0; }
        .content { padding: 20px; }
        
        /* Improved summary design */
        .summary { background: #f8fafc; border: 1px solid #e2e8f0; border-radius: 8px; padding: 20px; margin-bottom: 30px; }
        .summary-title { font-size: 16px; font-weight: 600; color: #374151; margin-bottom: 15px; }
        .stats-row { display: flex; align-items: center; gap: 20px; flex-wrap: wrap; }
        .stat-item { display: flex; align-items: center; gap: 8px; }
        .stat-item a { text-decoration: none; color: inherit; display: flex; align-items: center; gap: 8px; }
        .stat-item a:hover { opacity: 0.8; }
        .stat-icon { width: 16px; height: 16px; border-radius: 50%; }
        .stat-icon.passed { background: #10b981; }
        .stat-icon.failed { background: #ef4444; }
        .stat-icon.ignored { background: #f59e0b; }
        .stat-text { font-size: 14px; color: #374151; }
        .stat-number { font-weight: 600; }
        .duration { margin-left: auto; font-size: 14px; color: #6b7280; }
        
        .section { margin-bottom: 30px; background: #fafbfc; border: 1px solid #e5e7eb; border-radius: 8px; padding: 16px; }
        .section-title { font-size: 18px; font-weight: 600; color: #1e293b; }
        .section-title.collapsible { cursor: pointer; user-select: none; display: flex; justify-content: space-between; align-items: center; }
        .section-title.collapsible:hover { color: #374151; }
        .section-content { transition: max-height 0.3s ease; overflow: hidden; padding-top: 15px; }
        .section-content.collapsed { max-height: 0; padding-top: 0; }
        .section-title.collapsible .section-info { display: flex; align-items: center; gap: 8px; }
        .section-title.collapsible::after { content: '‚ñº'; font-size: 14px; transition: transform 0.2s; line-height: 1; }
        .section-title.collapsible.collapsed::after { transform: rotate(-90deg); }
        /* Improved clickable indication */
        .test-item { background: #f8fafc; border: 1px solid #e2e8f0; border-radius: 6px; margin-bottom: 10px; }
        .test-item:has(.test-details.show) { border-radius: 6px 6px 0 0; }
        .test-details { border-top: 1px solid #e2e8f0; display: none; margin: 0 -1px -1px -1px; background: #1e293b; border-radius: 0 0 6px 6px; }
        .test-header { padding: 15px; display: flex; justify-content: space-between; align-items: center; }
        .test-header.clickable { cursor: pointer; transition: background-color 0.2s; }
        .test-header.clickable:hover { background: #f1f5f9; }
        
        .test-right { display: flex; align-items: center; gap: 10px; }
        .view-output { font-size: 12px; color: #6b7280; transition: opacity 0.2s; }
        .view-output.no-output { opacity: 0.6; }
        .test-header:hover .view-output { color: #374151; }
        .test-header:hover .view-output.no-output { opacity: 1; }
        
        .test-name { font-family: 'Monaco', 'Menlo', monospace; font-size: 14px; }
        .test-status { padding: 4px 8px; border-radius: 4px; font-size: 12px; font-weight: 500; }
        .test-status.passed { background: #dcfce7; color: #166534; }
        .test-status.failed { background: #fecaca; color: #991b1b; }
        .test-status.ignored { background: #fef3c7; color: #92400e; }
        .test-details.show { display: block; }
        .test-output { color: #e2e8f0; padding: 15px; border-radius: 0; font-family: 'Monaco', 'Menlo', monospace; font-size: 13px; white-space: pre-wrap; overflow-x: auto; }
        .test-output a { color: #60a5fa; text-decoration: underline; }
        .test-output a:hover { color: #93c5fd; }
        .errors { background: #fef2f2; border: 1px solid #fecaca; border-radius: 6px; padding: 15px; }
        .error-item { margin-bottom: 10px; font-family: 'Monaco', 'Menlo', monospace; font-size: 13px; }
        .raw-lines { background: #f8fafc; border: 1px solid #e2e8f0; border-radius: 6px; padding: 15px; }
        .raw-line { font-family: 'Monaco', 'Menlo', monospace; font-size: 13px; margin-bottom: 5px; }
        
        @media (max-width: 600px) {
            .test-status { display: none; }
        }
    </style>
    <script>
        function toggleDetails(element) {
            const details = element.nextElementSibling;
            const wasVisible = details.classList.contains('show');
            details.classList.toggle('show');
            
            // Keep the clicked element in view if we're expanding
            if (!wasVisible) {
                element.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
            }
        }
        
        function toggleSection(element) {
            const content = element.nextElementSibling;
            const wasCollapsed = content.classList.contains('collapsed');
            content.classList.toggle('collapsed');
            element.classList.toggle('collapsed');
            
            // Keep the section header in view if we're expanding
            if (wasCollapsed) {
                element.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
            }
        }
    </script>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Cargo Test Report</h1>
        </div>
        <div class="content">
            <div class="summary">
                <div class="summary-title">Test Results Summary</div>
                <div class="stats-row">
                    <div class="stat-item">
                        <a href="#passed-tests">
                            <div class="stat-icon passed"></div>
                            <div class="stat-text"><span class="stat-number">1</span> passed</div>
                        </a>
                    </div>
                    <div class="stat-item">
                        <a href="#failed-tests">
                            <div class="stat-icon failed"></div>
                            <div class="stat-text"><span class="stat-number">0</span> failed</div>
                        </a>
                    </div>
                    <div class="stat-item">
                        <a href="#ignored-tests">
                            <div class="stat-icon ignored"></div>
                            <div class="stat-text"><span class="stat-number">0</span> ignored</div>
                        </a>
                    </div>
                    <div class="duration">‚è±Ô∏è 0.001s</div>
                </div>
            </div>


            <div class="section">
                <div class="section-title collapsible collapsed" id="passed-tests" onclick="toggleSection(this)">
                    <div class="section-info">
                        <span>‚úÖ Passed Tests</span>
                        <span>(1)</span>
                    </div>
                </div>
                <div class="section-content collapsed">
                <div class="test-item">
                    <div class="test-header clickable" onclick="toggleDetails(this)">
                        <span class="test-name">tests::simple_test</span>
                        <div class="test-right">
                            <span class="view-output">üëÅÔ∏è View output</span>
                            <span class="test-status passed">PASSED</span>
                        </div>
                    </div>
                    <div class="test-details">
                        <div class="test-output">test output
</div>
                    </div>
                </div>
                </div>
            </div>



            <div class="section">
                <div class="section-title collapsible collapsed" onclick="toggleSection(this)">
                    <div class="section-info">
                        <span>üìÑ Raw Output</span>
                    </div>
                </div>
                <div class="section-content collapsed">
                <div class="raw-lines">
                    <div class="raw-line">Compiling test-project v0.1.0</div>
                    <div class="raw-line">Finished test [unoptimized + debuginfo] target(s) in 0.12s</div>
                    <div class="raw-line">Running unittests src/lib.rs (target/debug/deps/test_project-abc123)</div>
                </div>
                </div>
            </div>
        </div>
    </div>
</body>
</html>
